File: 1744.can-you-eat-your-favorite-candy-on-your-favorite-day.py
#
# @lc app=leetcode id=1744 lang=python3
#
# [1744] Can You Eat Your Favorite Candy on Your Favorite Day?
#

# @lc code=start
import itertools
class Solution:
    def canEat(self, candiesCount: list[int], queries: list[list[int]]) -> list[bool]:
        # ans = []
        # preSum = [0] * (len(candiesCount)  +1)
        # for i in range(len(candiesCount)):
        #     preSum[i + 1] = preSum[i] + candiesCount[i]
        # for query in queries:
        #     need_lower = preSum[query[0]] + 1
        #     need_upper = preSum[query[0] + 1]
        #     lb = query[1] + 1
        #     ub = (query[1] + 1) * query[2]
        #     if need_upper < lb or need_lower > ub:
        #         ans.append(False)
        #     else:
        #         ans.append(True)
        # return ans   
        # 
        #todo Optimized solution from LeeCode
        A = [0] + list(itertools.accumulate(candiesCount))
        return [A[t] // c <= d < A[t + 1] for t, d, c in queries]
         
# @lc code=end

File: 1744.can-you-eat-your-favorite-candy-on-your-favorite-day.java
/*
 * @lc app=leetcode id=1744 lang=java
 *
 * [1744] Can You Eat Your Favorite Candy on Your Favorite Day?
 */

// @lc code=start
class Solution {
    public boolean[] canEat(int[] candiesCount, int[][] queries) {
        //1 <= amount eat per day <= dailyCap
        int n = candiesCount.length, m = queries.length;
        long[] preSum = new long[n + 1];
        boolean[] ans = new boolean[m];
        for(int i = 0; i < n; i++){
            preSum[i + 1] = preSum[i] + candiesCount[i];
        }
        for(int i = 0; i < m; i++){
            int favType = queries[i][0], favDay = queries[i][1], dailyCap = queries[i][2];
            Long min = (long) (favDay + 1)* 1;
            long max = (long) (favDay + 1)*dailyCap;
            long total = preSum[favType + 1];
            ans[i] = !(total < min || preSum[favType] >= max);
        }
        return ans;
    }
}
// @lc code=end

